From 947e01be9d055786b4f730ef0bcc5f53b5fdc9dd Mon Sep 17 00:00:00 2001
From: Subbaraya Sundeep <sbhatta@marvell.com>
Date: Mon, 27 Apr 2020 12:46:08 +0530
Subject: [PATCH 18/37] octeontx2-af: Fix MSIX handler mailbox

commit 1aa116e9cea3fd66265df81c6101890d3af738cd from
git@git.assembla.com:cavium/WindRiver.linux.git

Changes for 98xx in msix_offset mbox handler
returns error for NIX block if no NIX LF is attached
to the caller. Instead return error in message
response only since calling msix_offset before
attaching LFs is not invalid.

Fixes commit:
3bb5a5d9 ("octeontx2-af: Return assigned NIX/CPT block address")

Change-Id: I7ed4226b3fbd4ad6e60059d40a98d0135c23071b
Signed-off-by: Subbaraya Sundeep <sbhatta@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/27534
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/af/rvu.c | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu.c b/drivers/net/ethernet/marvell/octeontx2/af/rvu.c
index fc71835d0710..01b3d321961d 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu.c
@@ -1575,11 +1575,12 @@ int rvu_mbox_handler_msix_offset(struct rvu *rvu, struct msg_req *req,
 
 	/* Get BLKADDR from which LFs are attached to pcifunc */
 	blkaddr = rvu_get_blkaddr(rvu, BLKTYPE_NIX, pcifunc);
-	if (blkaddr < 0)
-		return blkaddr;
-
-	lf = rvu_get_lf(rvu, &hw->block[blkaddr], pcifunc, 0);
-	rsp->nix_msixoff = rvu_get_msix_offset(rvu, pfvf, blkaddr, lf);
+	if (blkaddr < 0) {
+		rsp->nix_msixoff = MSIX_VECTOR_INVALID;
+	} else {
+		lf = rvu_get_lf(rvu, &hw->block[blkaddr], pcifunc, 0);
+		rsp->nix_msixoff = rvu_get_msix_offset(rvu, pfvf, blkaddr, lf);
+	}
 
 	rsp->sso = pfvf->sso;
 	for (slot = 0; slot < rsp->sso; slot++) {
-- 
2.17.1

